source/Octopus.Tentacle/Services/Scripts/ScriptServiceV2.cs

```diff
@@ -0,0 +1,165 @@
+                var process = LaunchShell(command.ScriptTicket, command.TaskId, workspace, runningScript.ScriptStateStore, cancellationTokenSource);
+
+                runningScript.Process = process;
+                runningScript.CancellationTokenSource = cancellationTokenSource;
+
+                if (command.DurationToWaitForScriptToFinish > TimeSpan.Zero)
+                {
+                    var waited = Stopwatch.StartNew();
+                    while (process.State != ProcessState.Complete && waited.Elapsed < command.DurationToWaitForScriptToFinish)
+                    {
+                        Thread.Sleep(TimeSpan.FromMilliseconds(10));
+                    }
+                }
+
+                return GetResponse(command.ScriptTicket, 0, runningScript.Process);
+            }
+        }
+
+        public ScriptStatusResponseV2 GetStatus(ScriptStatusRequestV2 request)
+        {
+            runningScripts.TryGetValue(request.Ticket, out var runningScript);
+            return GetResponse(request.Ticket, request.LastLogSequence, runningScript?.Process);
+        }
+
+        public ScriptStatusResponseV2 CancelScript(CancelScriptCommandV2 command)
+        {
+            runningScripts.TryGetValue(command.Ticket, out var runningScript);
+
+            runningScript?.CancellationTokenSource?.Cancel();
+
+            return GetResponse(command.Ticket, command.LastLogSequence, runningScript?.Process);
+        }
+
+        public void CompleteScript(CompleteScriptCommandV2 command)
+        {
+            runningScripts.TryRemove(command.Ticket, out _);
+            var workspace = workspaceFactory.GetWorkspace(command.Ticket);
+            workspace.Delete();
+        }
+
+        RunningScript LaunchShell(ScriptTicket ticket, string serverTaskId, IScriptWorkspace workspace, IScriptStateStore stateStore, CancellationTokenSource cancel)
+        {
+            var runningScript = new RunningScript(shell, workspace, stateStore, workspace.CreateLog(), serverTaskId, cancel.Token, log);
+            var thread = new Thread(runningScript.Execute) { Name = "Executing PowerShell runningScript for " + ticket.TaskId };
+            thread.Start();
+            return runningScript;
+        }
+
+        ScriptStatusResponseV2 GetResponse(ScriptTicket ticket, long lastLogSequence, RunningScript? runningScript)
+        {
+            var workspace = workspaceFactory.GetWorkspace(ticket);
+            var scriptLog = runningScript?.ScriptLog ?? workspace.CreateLog();
+            var logs = scriptLog.GetOutput(lastLogSequence, out var next);
+
+            if (runningScript != null)
+            {
+                return new ScriptStatusResponseV2(ticket, runningScript.State, runningScript.ExitCode, logs, next);
+            }
+
+            // If we don't have a RunningProcess we check the ScriptStateStore to see if we have persisted a script result
+            var scriptStateStore = scriptStateStoreFactory.Get(ticket, workspace);
```

Comment:
`ScriptStateStoreFactory.Get` never returns null. Is there a reason to use `?` here?

Reply:
There is not. I suspect it's due to some significant refactoring during development. Good spot, will clean up 👍